nx/lib/nx/defn/evaluator.ex

```diff
@@ -42,135 +42,304 @@ defmodule Nx.Defn.Evaluator do
+        {res, [decrement_cache(cache, id, count, res) | caches]}
 
       %{} ->
-        {res, cache} = eval_apply(op, ans, state, cache)
+        # If we don't find the tensor in the current scope,
+        # it may be in a parent scope, so look up. If we find
+        # it in the parent, we don't decrement it, as that will
+        # be done at the end of processing the current scope.
+        eval_parent(caches, id, op, ans, state, [cache])
+    end
+  end
+
+  defp eval(other, _state, [_ | _] = caches) do
+    {other, caches}
+  end
+
+  defp decrement_cache(cache, id, 1, _res), do: Map.delete(cache, id)
+  defp decrement_cache(cache, id, counter, res), do: %{cache | id => {counter - 1, res}}
+
+  defp eval_parent([cache | caches], id, op, ans, state, acc) do
+    case cache do
+      %{^id => {_count, res}} ->
+        {res, Enum.reverse(acc, [cache | caches])}
+
+      %{^id => count} when is_integer(count) ->
+        {res, [cache | caches]} = eval_apply(op, ans, state, [cache | caches])
         state.gc && :erlang.garbage_collect(self())
-        {res, Map.put(cache, id, res)}
+        {res, Enum.reverse(acc, [Map.put(cache, id, {count, res}) | caches])}
+
+      %{} ->
+        eval_parent(caches, id, op, ans, state, [cache | acc])
     end
   end
 
-  defp eval(other, _state, cache) do
-    {other, cache}
+  defp eval_parent([], _id, _op, ans, _state, _acc) do
+    raise "trying to read evaluator cache that has expired during expression:\n\n#{inspect(ans)}\n\n" <>
+            "Please report this bug with the relevant code that triggers it: https://github.com/elixir-nx/nx"
   end
 
-  defp eval_apply(:parameter, %{data: %Expr{args: [i]}}, state, cache) do
+  defp decrement_parents([cache | caches], id) do
+    case cache do
+      %{^id => {count, value}} -> [decrement_cache(cache, id, count, value) | caches]
+      %{^id => count} -> [%{cache | id => count - 1} | caches]
+      %{} -> [cache | decrement_parents(caches, id)]
+    end
+  end
+
+  defp eval_apply(:parameter, %{data: %Expr{args: [i]}}, state, caches) do
     case Enum.fetch!(state.params, i).() do
       %Nx.Tensor{data: %Nx.Defn.Expr{}} = tensor ->
         raise ArgumentError,
               "cannot pass a tensor expression as argument to defn, got: #{inspect(tensor)}"
 
       %Nx.Tensor{} = tensor ->
-        {tensor, cache}
+        {tensor, caches}
     end
   end
 
-  defp eval_apply(:fun, %{data: %Expr{args: [args, expr, _mfa]}}, state, cache) do
+  defp eval_apply(:elem, %Nx.Tensor{data: %Expr{args: [tuple, i]}}, state, caches) do
+    {tuple, caches} = composite_eval(tuple, state, caches)
+    {elem(tuple, i), caches}
+  end
+
+  defp eval_apply(:attach_token, %Nx.Tensor{data: %Expr{args: [token, expr]}}, state, caches) do
+    {_, caches} = eval(token, state, caches)
+    eval(expr, state, caches)
+  end
+
+  defp eval_apply(:fun, %{data: %Expr{args: [args, expr, _mfa], id: id}}, state, caches) do
+    fun_cache = Map.fetch!(hd(caches), [:fun | id])
```

Comment:
Curious: what's the benefit over `{:fun, id}`?

Reply:
It is a (bad?) habit I picked: `[:fun | id]` uses one word fewer compared to tuple, but honestly unlikely to matter here. :)

Reply:
Good to know, thanks :) For a moment it made me wonder if `id` isn't a list, it probably sacrifices some readability.